File: contrib/go/bareagent.go
// Package bareagent provides a thin subprocess wrapper for bare-agent.
// Communicates via JSONL over stdin/stdout. Zero external dependencies.
package bareagent

import (
	"bufio"
	"encoding/json"
	"fmt"
	"io"
	"os/exec"
)

// Result holds the agent's response.
type Result struct {
	Text      string         `json:"text"`
	ToolCalls []any          `json:"toolCalls"`
	Usage     map[string]int `json:"usage"`
	Error     string         `json:"error,omitempty"`
}

// Event holds a JSONL event from the agent.
type Event struct {
	Type string         `json:"type"`
	Data map[string]any `json:"data"`
	Ts   string         `json:"ts"`
}

// BareAgent wraps a bare-agent subprocess.
type BareAgent struct {
	cmd     *exec.Cmd
	stdin   io.WriteCloser
	scanner *bufio.Scanner
	OnEvent func(Event)
}

// New spawns a bare-agent subprocess with the given provider and options.
func New(provider, model, url string) (*BareAgent, error) {
	args := []string{"bare-agent", "--jsonl", "--provider", provider}
	if model != "" {
		args = append(args, "--model", model)
	}
	if url != "" {
		args = append(args, "--url", url)
	}
	cmd := exec.Command("npx", args...)
	stdin, err := cmd.StdinPipe()
	if err != nil {
		return nil, fmt.Errorf("stdin pipe: %w", err)
	}
	stdout, err := cmd.StdoutPipe()
	if err != nil {
		return nil, fmt.Errorf("stdout pipe: %w", err)
	}
	if err := cmd.Start(); err != nil {
		return nil, fmt.Errorf("start: %w", err)
	}
	return &BareAgent{
		cmd:     cmd,
		stdin:   stdin,
		scanner: bufio.NewScanner(stdout),
	}, nil
}

// Run sends a goal string or messages and blocks until a result or error event.
func (a *BareAgent) Run(goal string) (*Result, error) {
	req := map[string]any{
		"method": "run",
		"params": map[string]any{"goal": goal},
	}
	data, err := json.Marshal(req)
	if err != nil {
		return nil, err
	}
	if _, err := fmt.Fprintf(a.stdin, "%s\n", data); err != nil {
		return nil, fmt.Errorf("write stdin: %w", err)
	}
	for a.scanner.Scan() {
		var event Event
		if err := json.Unmarshal(a.scanner.Bytes(), &event); err != nil {
			continue
		}
		if a.OnEvent != nil && event.Type != "result" && event.Type != "error" {
			a.OnEvent(event)
		}
		if event.Type == "result" || event.Type == "error" {
			var result Result
			raw, _ := json.Marshal(event.Data)
			json.Unmarshal(raw, &result)
			return &result, nil
		}
	}
	return nil, fmt.Errorf("subprocess closed without result")
}

// RunMessages sends a messages array and blocks until a result.
func (a *BareAgent) RunMessages(messages []map[string]string) (*Result, error) {
	req := map[string]any{
		"method": "run",
		"params": map[string]any{"messages": messages},
	}
	data, err := json.Marshal(req)
	if err != nil {
		return nil, err
	}
	if _, err := fmt.Fprintf(a.stdin, "%s\n", data); err != nil {
		return nil, fmt.Errorf("write stdin: %w", err)
	}
	for a.scanner.Scan() {
		var event Event
		if err := json.Unmarshal(a.scanner.Bytes(), &event); err != nil {
			continue
		}
		if a.OnEvent != nil && event.Type != "result" && event.Type != "error" {
			a.OnEvent(event)
		}
		if event.Type == "result" || event.Type == "error" {
			var result Result
			raw, _ := json.Marshal(event.Data)
			json.Unmarshal(raw, &result)
			return &result, nil
		}
	}
	return nil, fmt.Errorf("subprocess closed without result")
}

// Close terminates the subprocess.
func (a *BareAgent) Close() error {
	a.stdin.Close()
	return a.cmd.Wait()
}

File: contrib/java/BareAgent.java
import java.io.*;
import java.util.*;

/**
 * bare-agent Java wrapper -- subprocess + JSONL over stdin/stdout.
 * stdlib only: ProcessBuilder, simple JSON parsing (no javax.json needed).
 *
 * Usage:
 *   BareAgent agent = new BareAgent("openai", "gpt-4o-mini", null);
 *   Map<String, Object> result = agent.run("What is 2 + 2?");
 *   System.out.println(result.get("text"));
 *   agent.close();
 */
public class BareAgent {

    private final Process process;
    private final BufferedWriter stdin;
    private final BufferedReader stdout;
    private EventCallback onEvent;

    @FunctionalInterface
    public interface EventCallback {
        void onEvent(String type, String rawJson);
    }

    public BareAgent(String provider, String model, String url) throws IOException {
        List<String> cmd = new ArrayList<>(Arrays.asList(
            "npx", "bare-agent", "--jsonl", "--provider", provider
        ));
        if (model != null) { cmd.add("--model"); cmd.add(model); }
        if (url != null) { cmd.add("--url"); cmd.add(url); }

        ProcessBuilder pb = new ProcessBuilder(cmd);
        pb.redirectErrorStream(false);
        pb.environment().putAll(System.getenv());
        this.process = pb.start();
        this.stdin = new BufferedWriter(new OutputStreamWriter(process.getOutputStream()));
        this.stdout = new BufferedReader(new InputStreamReader(process.getInputStream()));
    }

    public void setOnEvent(EventCallback callback) {
        this.onEvent = callback;
    }

    /** Send a goal string and block until result. */
    public String run(String goal) throws IOException {
        String req = "{\"method\":\"run\",\"params\":{\"goal\":" + jsonEscape(goal) + "}}";
        stdin.write(req);
        stdin.newLine();
        stdin.flush();
        return readUntilTerminal();
    }

    /** Send a pre-built JSON params string and block until result. */
    public String runRaw(String paramsJson) throws IOException {
        String req = "{\"method\":\"run\",\"params\":" + paramsJson + "}";
        stdin.write(req);
        stdin.newLine();
        stdin.flush();
        return readUntilTerminal();
    }

    private String readUntilTerminal() throws IOException {
        String line;
        while ((line = stdout.readLine()) != null) {
            line = line.trim();
            if (line.isEmpty()) continue;
            String type = extractType(line);
            if (type != null && !type.equals("result") && !type.equals("error")) {
                if (onEvent != null) onEvent.onEvent(type, line);
                continue;
            }
            if ("result".equals(type) || "error".equals(type)) {
                return extractData(line);
            }
        }
        return null;
    }

    /** Terminate the subprocess. */
    public void close() {
        try { stdin.close(); } catch (IOException ignored) {}
        try { process.waitFor(); } catch (InterruptedException ignored) {}
    }

    // Minimal JSON helpers -- avoids external deps.

    private static String jsonEscape(String s) {
        StringBuilder sb = new StringBuilder("\"");
        for (char c : s.toCharArray()) {
            switch (c) {
                case '"':  sb.append("\\\""); break;
                case '\\': sb.append("\\\\"); break;
                case '\n': sb.append("\\n"); break;
                case '\r': sb.append("\\r"); break;
                case '\t': sb.append("\\t"); break;
                default:   sb.append(c);
            }
        }
        return sb.append("\"").toString();
    }

    private static String extractType(String json) {
        int i = json.indexOf("\"type\"");
        if (i < 0) return null;
        int colon = json.indexOf(':', i + 6);
        if (colon < 0) return null;
        int q1 = json.indexOf('"', colon + 1);
        if (q1 < 0) return null;
        int q2 = json.indexOf('"', q1 + 1);
        if (q2 < 0) return null;
        return json.substring(q1 + 1, q2);
    }

    private static String extractData(String json) {
        int i = json.indexOf("\"data\"");
        if (i < 0) return json;
        int colon = json.indexOf(':', i + 6);
        if (colon < 0) return json;
        // Return everything from after the colon to before the last }
        // This is a rough extraction -- for production use, add a JSON library
        int start = colon + 1;
        while (start < json.length() && json.charAt(start) == ' ') start++;
        int end = json.lastIndexOf('}');
        if (end <= start) return json;
        return json.substring(start, end);
    }

    public static void main(String[] args) throws Exception {
        BareAgent agent = new BareAgent("openai", "gpt-4o-mini", null);
        System.out.println("Spawned bare-agent subprocess");
        agent.setOnEvent((type, raw) ->
            System.out.printf("[%s] %s%n", type, raw)
        );
        String result = agent.run("What is 2 + 2?");
        System.out.println("Result: " + result);
        agent.close();
        System.out.println("Done");
    }
}
